### ISSUE TITLE
[DISCUSS] Notebook actions while the notebook is not yet fully rendered by the virtual notebook

### ISSUE THREAD

**Issue by echarles** [jupyterlab/jupyterlab#10933]
This is a followup of https://github.com/jupyterlab/jupyterlab/issues/10840

I looks also that the reported case https://github.com/jupyterlab/jupyterlab/issues/10840#issuecomment-903130229 still gives failures

The initial idea of the virtual notebook was to have 2 cell "areas":

1. The first one at the top of the notebook which are rendered directly.
2. The second one after the first one, which are populated by placeholders, and then lazily rendered when the browser is idle.
We were using in the second area an index which has shown to be broken after #10793. This PR fixes that.

Still, there is the assumption that during the rendering of the second area, the user will not take actions to change the structure of the notebook. It is fine for him to run an existing cell, write in existing cell, save the notebook, restart the kernel... but I can see an issue if he inserts or deletes a cell, as the position of the cells to be rendered should be changed. We have 2 options to takes this behavior into account:

1. Disallow any structural change to the cells indices while the notebook is not fully rendered... in other words the "toRenderCell" map is empty
2. On insert/delete, update the indices of the "toRenderMap" according to the change, which is possible, but a bit cumbersome to write in code.

@krassowski made goof points on https://github.com/jupyterlab/jupyterlab/pull/10898#issuecomment-904416206 which I copy here after.

In my opinion if we will have a visual cue of loading, this could be an acceptable UX (barely, but acceptable). On the other hand, I am not convinced whether making good visual cues and prohibiting changes from triggering any susceptible code path will be easier than tracking the changes. I mean that a good UX would only prohibit very specific actions and only on a subset of cells: by the time we add the logic to handle all those actions we already covered the same places as we would if we instead implemented the index tracking/rebuilding (which is obviously more difficult in itself, but the extend places in code that need to be touched will not be so much different). Can you raise this tomorrow on the dev meeting (unfortunately I cannot attend it tomorrow)?

**Comment by Mithil467** [comment]
#10998 has an example that still causes this issue on latest master.
<details>
<summary>ipynb file</summary>
<pre>
{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e93a11d0-9c02-4550-812e-b88208da40e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d8bddb3b-e5e6-4820-b522-f32590545333",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1207dcec-fafc-4ef8-b8bd-df50889e13b6",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff9babe7-0f5e-45fb-98b7-c96cce1497f2",
   "metadata": {},
   "source": [
    "# 特征"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "aa741a5c-58d2-455d-93e0-bb35712c8f9f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b0b10f7-97d2-4ed0-abe2-0795edae044d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "779fd2ce-7b4e-42b4-87ea-280a1598f664",
   "metadata": {},
   "source": [
    "# 数据拉取"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e77c90b9-e061-437e-b747-5938b6981472",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "58642a3a-ba1f-4f08-a7c0-9355eecbe141",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cdb62db1-437d-462b-b993-1c60b04df233",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "77936c58-883f-440b-84c7-e34d27e331e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ab373a4-47c8-4557-901e-9f0b6136adaf",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d10ee40-9a5e-45c9-bf70-f9e68826f6a0",
   "metadata": {},
   "source": [
    "# 数据预处理"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f275d75e-9ab4-413d-914b-811d6003e64c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9bac9859-a0a5-4dfc-876b-2f590ba30cbb",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "002930ab-67f8-4dc8-9d7b-c32ecda8ebb2",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a5dadccd-b3f1-4f2d-8cd9-ea108425b413",
   "metadata": {},
   "source": [
    "# xgb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bb43e452-f6dd-4370-8bb4-41332666ba58",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c9c248f4-674b-4bf6-b019-53f54b5e03fb",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f9e3633-59b4-4017-bb42-53b26d56f66e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cb32f18a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_pickle('./dnn_mob3_data_xgb.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e223df-0478-45c5-bad7-8d3743291f6a",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 训练时模型结构"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5df464e4-d134-4d42-9131-46615252cb4e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3a116667-cc43-4e7b-9b41-f51a94307877",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59bf278d-f33e-4844-8fc8-23991ac99c22",
   "metadata": {},
   "source": [
    "# 剔除辅助任务"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "46582a77-1296-4a13-ac95-21ac12879b3b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "697c8842-61c9-403e-bfa1-5e888b57160a",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "18fdcad5-da3f-43d1-a2a9-9b81dd790072",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11a11b30",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1fd2d92e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8699504a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5d0d9225",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87053e25-eebf-4ed9-9464-8a4480e3521b",
   "metadata": {},
   "source": [
    "# pb文件"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16a13ed8-74de-4f23-89cd-4cf60a6d960a",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7bcdf186-20f8-4e73-841a-e2715cf6ac0c",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6d69c019-680e-46c1-9b17-8793c37ebc63",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cc2f31c-8395-40a4-a13f-94c75d205251",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2798f9ee-e6cb-4ab2-98a2-41d1c5af2482",
   "metadata": {
    "tags": []
   },
   "source": [
    "# ks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a75342b0-f313-4ff6-bd9b-31ced0d59b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f584a7c-bb66-4b03-a0b1-7ac01c89868d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a771b33-aab9-4c77-9c00-cc2d71b132dd",
   "metadata": {},
   "source": [
    "# 交叉矩阵"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6c27a03-ff47-48f7-af6e-75ca783d2491",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68e9588e",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327a3b63-d86b-4607-8cfd-e43fcea9d16f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3feaf88-ad46-4459-89c6-403c3e55b1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c6c60f-371e-42bb-a20a-bc6752417822",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db033789-4957-4a09-a62c-4dcae00ee47b",
   "metadata": {},
   "source": [
    "# top20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a7669540",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bde9e530",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5a4254c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "dd2d23ab",
   "metadata": {},
   "source": [
    "# 融合"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4845f01",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fc186cd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6f846863",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e46f9ae7-7ce7-4200-ab70-d9ffaff34e88",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8299eea1-d103-4a84-9ca1-5baa3cb286ca",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67eb04bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "13941610",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7a9a654d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ea97bcd3",
   "metadata": {},
   "source": [
    "# 特征"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28e2f285",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5da8680",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29024b44",
   "metadata": {},
   "source": [
    "# 数据拉取"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28b03355",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d1a9bcce",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4d6344bc",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1b56f1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35b7323b",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd909e50",
   "metadata": {},
   "source": [
    "# 数据预处理"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cca06018",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d91d2638",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8b63ed53",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a7bb1f56",
   "metadata": {},
   "source": [
    "# xgb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a24dbf4",
   "metadata": {
    "tags": []
   },
   "source": [
    "# 训练时模型结构"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "20a6e4fe",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff822904",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4efea718",
   "metadata": {},
   "source": [
    "# 剔除辅助任务"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60248327",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "22d1660b",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_model.summary()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "796cc7fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2798f9ee-e6cb-4ab2-98a2-41d1c5af2482",
   "metadata": {
    "tags": []
   },
   "source": [
    "# ks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a75342b0-f313-4ff6-bd9b-31ced0d59b29",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f584a7c-bb66-4b03-a0b1-7ac01c89868d",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1a771b33-aab9-4c77-9c00-cc2d71b132dd",
   "metadata": {},
   "source": [
    "# 交叉矩阵"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6c27a03-ff47-48f7-af6e-75ca783d2491",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "327a3b63-d86b-4607-8cfd-e43fcea9d16f",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f3feaf88-ad46-4459-89c6-403c3e55b1c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "15c6c60f-371e-42bb-a20a-bc6752417822",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db033789-4957-4a09-a62c-4dcae00ee47b",
   "metadata": {},
   "source": [
    "# top20"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e46f9ae7-7ce7-4200-ab70-d9ffaff34e88",
   "metadata": {},
   "outputs": [],
   "source": [
    "print(\"a\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8299eea1-d103-4a84-9ca1-5baa3cb286ca",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
</pre>
</details>

**Comment by echarles** [comment]
Thx @Mithil467 for reporting and providing a reproducer. The given notebook has duplicate cell ids, which generate exceptions in the virtual rendering. Have you forged it by hand?

Playing devil's advocate, I would say that this is a feature rather than a bug.... A "malformed" notebook (need to check if nbformat imposes unique ids, but if it does not, I would say it should) should not be rendered without at least informing the user that the content has an issue. Disabling the virtual notebook will render it correctly, but this may be tagged as a bug :)

Seriously, I guess this is a very edge case that should be better managed, showing to the user a nice message. Thinking loud, I don't see an easy way with the current virtual notebook, to support non-unique cell ids.

**Comment by Mithil467** [comment]
I got this reproducer from this comment here https://github.com/jupyterlab/jupyterlab/issues/10998#issuecomment-913611317. Otherwise I don't know much about its origin or if it was forged in any way, but that has to be the case.

**Comment by consideRatio** [comment]
I'm not sure why I ended up with a notebook with duplicate cell ids, perhaps a bug that has since been fixed?

Anyhow, I had such notebook and I observed horizontal line artifacts, so I reduced its complexity to a minimal example of how duplicate cell IDs can lead to these artifacts to show in the latest current version of jupyterlab (3.2.9).

![image](https://user-images.githubusercontent.com/3837114/155782555-4c715eee-89e1-433a-b0a0-b1b8d676a6fb.png)

<details>
  <summary>Notebook above in .json</summary>

```json
{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-0",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 0 (adding cells with unique ids increases the number of line artifacts)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-1",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-2",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-2",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 2 - duplication 1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-2",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 2 - duplication 2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-2",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 2 - duplication 3"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-2",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 2 - duplication 4 (adding cells with duplicate ids also increases the number of line artifacts)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-3",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 3"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4 - duplication 1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4 - duplication 2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4 - duplication 3"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4 - duplication 4"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4 - duplication 5"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-4",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 4 - duplication 6"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 1"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 2"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 3"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 4"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 5"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 6"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 7"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 8"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 9"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 10"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 11"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-5",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 5 - duplication 12"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "id": "id-6",
      "metadata": {
        "tags": []
      },
      "outputs": [],
      "source": [
        "# 6 (adding cells down here didn't increases the number of line artifacts above)"
      ]
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "language": "python",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.8.6"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}
```

</details>

**Comment by fperez** [comment]
I wanted to flag this as a pretty serious issue - in the last few days, on two separate hubs (a UC Berkeley one and our JMTE one), I've also encountered this. Notebooks that were perfectly fine and I was working on seem to get randomly corrupted, with a lot of cells with duplicate IDs getting injected into them.

The resulting notebook looks like this:

![image](https://user-images.githubusercontent.com/57394/155822790-285a343d-c7e2-46a7-a7ce-24011fcd99b9.png)

Those horizontal lines correspond to corrupted cells with duplicate IDs that can't 

If I open that same notebook in Classic, it actually throws a validation error (Lab doesn't, it probably should):

![image](https://user-images.githubusercontent.com/57394/155822977-031b4212-3eec-4979-b5b5-caf696c9387b.png)

The same error gets thrown when I try to save in Classic, btw.

Now, in addition to those duplicate cell IDs, the notebook itself is actually duplicated in content - I think that's the origin of the problem. It's confusing b/c Lab renders those weird lines early on, but then the notebook has 100% of its content appended back onto itself at the end, so the whole thing is basically duplicated.

That duplication/corruption has happened to me in the last few weeks with at least 3 notebooks, though I can't tell what actions led to it.  But I'd say this should be super high priority - this is actual data corruption of working notebooks.

All this was on 3.2.9, running on two separate cloud-hosted hubs with JupyterHub, FWIW.

**Comment by fperez** [comment]
FYI - [this notebook](https://github.com/UCB-stat-159-s22/site/blob/main/lectures/climate-data.ipynb) causes the problem for me, but _not deterministically_. I've opened several times, doing a full `git reset --hard` each time to go  back to the original, and sometimes it gets duplicated/corrupted, sometimes not.

I'm happy to help try to track this down, it's become a pretty critical issue...

**Comment by echarles** [comment]
@fperez @consideRatio Thx for reporting. Can you confirm that the horizontal lines shown are notebooks with duplicates (you an open in classic notebook and see if the error message is shown). I just want to make sure that there is no issue with  virtual notebook rendering. 

A question is what to do with notebook with same ids? Are you OK to just not show it and just display similar message as the classic notebook one?

> Notebooks that were perfectly fine and I was working on seem to get randomly corrupted, with a lot of cells with duplicate IDs getting injected into them.

Another question is why suddenly notebooks get randomly corrupted?
Which version are you using? Have you upgraded recently? Any common pattern for those notebook contents, or usage pattern?

**Comment by fperez** [comment]
As for notebooks with duplicate cell IDs, I think there's two aspects:

1. It shouldn't happen from within the UI. In this case, it's really bad: _every_ cell gets duplicated, not just one or two. So a long notebook can end up being very big, and with many duplicate cells. That's a serious bug, period.
2. What the UI offers to do if it ever encounters a duplicate cell ID is a different matter. Other than this kind of corruption caused by us, it should be very rare, and it might be OK to simply deduplicate the ID by reassigning a new one, while obviously warning the user.  It could always provide an option to ask "delete duplicates or deduplicate IDs", or similar... But obviously if it's something very rare, it's not worth investing too much into a complex UI... A simple warn-and-deduplicate might be enough.

The real problem is #1 - that we are creating these massive duplications in the first place...

**Comment by echarles** [comment]
We can work for fix to show the error message in case of duplicate (eg. stopping rendering, replacing the duplicates by new ones...).

We should be more worried on the sudden corruption of notebooks. If you are pretty sure that it is further the upgrade, it will help us to look where the issue is.

For such case, a replicator is more than welcome. I can play with [https://github.com/UCB-stat-159-s22/site/blob/main/lectures/climate-data.ipynb], and hack to warn on duplicate id.

BTW from which version did you upgrade?

**Comment by echarles** [comment]
As a temporary urgent mitigation, is a downgrade an option for you? This would help also to assess the issue is in that specific `3.2.9` version.

**Comment by fperez** [comment]
I haven't been able to replicate it with 3.2.9 locally, but I'll keep trying. I can indeed downgrade the hub as well - the problem isn't fully deterministic though, so it will take a bit to clarify if this really is the cause or not. I'll keep reporting, and if anyone else has ideas please let us know.
